src/twisted/spread/test/test_pb.py

```diff
@@ -1434,19 +1511,34 @@ def test_loginLogout(self):
         # ignore it.
         mind = "BRAINS!"
 
-        d = factory.login(creds, mind)
+        loginCompleted = Deferred()
+
+        d = self.clientFactory.login(creds, mind)
         def cbLogin(perspective):
             self.assertTrue(self.realm.lastPerspective.loggedIn)
             self.assertIsInstance(perspective, pb.RemoteReference)
-            return self._disconnect(None, factory)
+            return loginCompleted
+
+        def cbDisconnect(ignored):
+            self.clientFactory.disconnect()
+            self.completeClientLostConnection()
+
         d.addCallback(cbLogin)
+        d.addCallback(cbDisconnect)
 
         def cbLogout(ignored):
             self.assertTrue(self.realm.lastPerspective.loggedOut)
         d.addCallback(cbLogout)
 
-        connector = reactor.connectTCP("127.0.0.1", self.portno, factory)
-        self.addCleanup(connector.disconnect)
+        self.establishClientAndServer()
+        self.pump.flush()
+        # The perspective passed to cbLogin has gone out of scope.
+        # Ensure its __del__ runs...
+        gc.collect()
+        # ...and send its decref message to the server
+        self.pump.flush()
+        # Now allow the client to disconnect.
+        loginCompleted.callback(None)
         return d
```

Comment:
Switching over to `SynchronousTestCase` exposes a bug in PB's error handling.  The [gory details](http://twistedmatrix.com/trac/ticket/8792#comment:5) are in Trac, but it suffices to say that fixing PB's error handling deserves [its own ticket](http://twistedmatrix.com/trac/ticket/8803).

In the meantime, is it OK to continue to use an asynchronous test case that returns `Deferred`s?  Another alternative would be to write a local helper that acts like `t.t_async.TestCase.assertFailure` and use it instead of `failureResultOf`, but it seems like a bad idea to heap a temporary work around into this PR.